### PR TITLE
Add support for NTLMv2

### PR DESCRIPTION
**Pull request by sboutzen**
Fix for https://github.com/denisenkom/go-mssqldb/issues/146

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -4,5 +4,6 @@ go 1.11
 
 require (
 	github.com/golang-sql/civil v0.0.0-20190719163853-cb61b32ac6fe
+	github.com/stretchr/testify v1.4.0
```
Comment (kardianos) [inline go.mod]: Please don't add a testing module. Please just use vanilla testing.T.

File: ntlm.go [hunk 2]
```diff
@@ -86,6 +89,20 @@ func utf16le(val string) []byte {
 	return v
 }
 
+func decodeUTF16le(val []byte) (string, error) {
```
Comment (kardianos) [inline ntlm.go]: The function you want is ucs22str I think. If this is different, then add a comment explaining the differences.

File: ntlm.go [hunk 3]
```diff
@@ -198,6 +217,84 @@ func ntlmSessionResponse(clientNonce [8]byte, serverChallenge [8]byte, password
...
+	hmacEntity := hmac.New(md5.New, passwordHash)
+	hmacEntity.Write(data)
+
+	return hmacEntity.Sum(nil)
+}
+
+func getTargetInformationData(bytes []byte) (infoBytes []byte, target string, err error) {
+	type2MessageLength := len(bytes)
+	targetNameAllocated := binary.LittleEndian.Uint16(bytes[14:16])
+	targetNameOffset := binary.LittleEndian.Uint32(bytes[16:20])
```
Comment (kardianos) [inline ntlm.go]: Please ensure that the type2MessageLength is > 20 before addressing that size.

File: ntlm.go [hunk 4]
```diff
@@ -198,6 +217,84 @@ func ntlmSessionResponse(clientNonce [8]byte, serverChallenge [8]byte, password
...
+	if type2MessageLength < int(targetNameOffset + uint32(targetNameAllocated)) {
+		return nil, target, errorInvalidServerType2Message
+	}
+
+	target, err = decodeUTF16le(bytes[targetNameOffset:targetNameOffset+uint32(targetNameAllocated)])
```
Comment (kardianos) [inline ntlm.go]: ucs22str

File: ntlm.go [hunk 5]
```diff
@@ -198,6 +217,84 @@ func ntlmSessionResponse(clientNonce [8]byte, serverChallenge [8]byte, password
...
+}
+
+func hmacMD5(passwordHash, data []byte) []byte {
+	hmacEntity := hmac.New(md5.New, passwordHash)
+	hmacEntity.Write(data)
+
+	return hmacEntity.Sum(nil)
+}
+
+func getTargetInformationData(bytes []byte) (infoBytes []byte, target string, err error) {
```
Comment (kardianos) [inline ntlm.go]: rename func to something more descriptive. No "data" include subject "ntlmv2" or similar or even more specific.
Don't make var name "bytes". You could call it payload, or even data, but not bytes.

Don't call that infoBytes, just info would be fine.
A comment on some of these messages would be helpful, especially references to a NTLMv2 specification.

File: ntlm.go [hunk 6]
```diff
@@ -198,6 +217,84 @@ func ntlmSessionResponse(clientNonce [8]byte, serverChallenge [8]byte, password
...
+	type2MessageLength := len(bytes)
+	targetNameAllocated := binary.LittleEndian.Uint16(bytes[14:16])
+	targetNameOffset := binary.LittleEndian.Uint32(bytes[16:20])
+	if type2MessageLength < int(targetNameOffset + uint32(targetNameAllocated)) {
+		return nil, target, errorInvalidServerType2Message
```
Comment (kardianos) [inline ntlm.go]: I'd prefer you put in a more specific error message here:
fmt.Errorf("mssql: NTLMv2 while parsing X message length %d too small for target offset %d", x, y)

File: ntlm.go [hunk 7]
```diff
@@ -198,6 +217,84 @@ func ntlmSessionResponse(clientNonce [8]byte, serverChallenge [8]byte, password
...
+
+	target, err = decodeUTF16le(bytes[targetNameOffset:targetNameOffset+uint32(targetNameAllocated)])
+	if err != nil {
+		return nil, target, err
+	}
+
+	targetInformationAllocated := binary.LittleEndian.Uint16(bytes[42:44])
+	targetInformationDataOffset := binary.LittleEndian.Uint32(bytes[44:48])
+	if type2MessageLength < int(targetInformationDataOffset + uint32(targetInformationAllocated)) {
+		return nil, target, errorInvalidServerType2Message
```
Comment (kardianos) [inline ntlm.go]: More specific error message please.

File: ntlm.go [hunk 8]
```diff
@@ -198,6 +217,84 @@ func ntlmSessionResponse(clientNonce [8]byte, serverChallenge [8]byte, password
...
+	}
+
+	targetInformationBytes := make([]byte, targetInformationAllocated)
+	copy(targetInformationBytes, bytes[targetInformationDataOffset:targetInformationDataOffset+uint32(targetInformationAllocated)])
+
+	return targetInformationBytes, target, nil
+}
+
+func getNTLMV2Response(target, username, password string, challenge, nonce [8]byte, targetInformation []byte, timestamp time.Time) []byte {
```
Comment (kardianos) [inline ntlm.go]: A bit of docs on this function would be great. Especially references to a spec. (no need to add docs within the function, but I have no way to know what it is expected to do right now)

File: ntlm.go [hunk 9]
```diff
@@ -198,6 +217,84 @@ func ntlmSessionResponse(clientNonce [8]byte, serverChallenge [8]byte, password
...
+	challengeAndBlob := make([]byte, challengeLength + blobLength)
+	copy(challengeAndBlob[:challengeLength], challenge[:])
+	copy(challengeAndBlob[challengeLength:], blob)
+	finalKey := hmacMD5(ntlmV2Hash, challengeAndBlob)
+	response := append(finalKey, blob...)
+
+	return response
+}
+
+func getLMV2Response(target, username, password string, challenge, nonce [8]byte) []byte {
```
Comment (kardianos) [inline ntlm.go]: func level doc with reference to spec would be appreciated.

File: ntlm.go [hunk 10]
```diff
@@ -212,11 +309,22 @@ func (auth *ntlmAuth) NextBytes(bytes []byte) ([]byte, error) {
...
-		nt_bytes := ntlmSessionResponse(nonce, challenge, auth.Password)
-		nt = nt_bytes[:]
+		if (flags & _NEGOTIATE_TARGET_INFO) != 0 {
+			timestamp := time.Now()
+			targetInformationBytes, target, err := getTargetInformationData(bytes)
+			if err != nil {
+				return nil, err
+			}
+
+			nt = getNTLMV2Response(target, auth.UserName, auth.Password, challenge, nonce, targetInformationBytes, timestamp)
```
Comment (kardianos) [inline ntlm.go]: I understand you are trying to encapsulate here, but getNTLMV2Response and getLMV2Response takes many of the same inputs, and does many of the same things.

Wouldn't it make sense to have one function that returned both the NTLM and LM response at the same time?

File: ntlm.go [hunk 11]
```diff
@@ -212,11 +309,22 @@ func (auth *ntlmAuth) NextBytes(bytes []byte) ([]byte, error) {
 	var lm, nt []byte
 	if (flags & _NEGOTIATE_EXTENDED_SESSIONSECURITY) != 0 {
```
Comment (kardianos) [inline ntlm.go]: Starting here, we see the server wants to neg session security, and we want to populate lm and nt messages.

If we pulled this out into a separate function, we can use a pattern of early returns to remove these nested if/elses, which makes reading harder. Now that you are adding a nested if/else, please try to do so.

File: ntlm_test.go [hunk 12]
```diff
@@ -3,8 +3,11 @@
 package mssql
 
 import (
+	"bytes"
 	"encoding/hex"
+	"github.com/stretchr/testify/assert"
```
Comment (kardianos) [inline ntlm_test.go]: Remove "testify/assert"

File: ntlm_test.go [hunk 13]
```diff
@@ -74,3 +77,38 @@ func TestNTLMSessionResponse(t *testing.T) {
...
+	lmV2Response := getLMV2Response(target, username, password, challenge, nonce)
+	if bytes.Compare(lmV2Response, expectedLMV2Response) != 0 {
+		t.Errorf("got:\n%s\nexpected:\n%s", hex.Dump(ntlmV2Response), hex.Dump(expectedNTLMV2Response))
+	}
+}
+
+func TestGetTargetInformationBlock(t *testing.T) {
+	bytes, _ := hex.DecodeString("4e544c4d53535000020000000600060038000000058289026999bc21067c77f40000000000000000ac00ac003e0000000a0039380000000f4600570042000200060046005700420001000c00590037004100410041003400040022006000700065002e00610058006e0071006e0070006e00650074002e0063006f006d00030030007900370041004100410034002e006000700065002e00610058006e0071006e0070006e00650074002e0063006f006d00050024006100610058006d002e00610058006e0071006e0070006e00650074002e0063006f006d00070008007d9647e8aed6d50100000000")
+	info, target, err := getTargetInformationData(bytes)
+	assert.Nil(t, err)
```
Comment (kardianos) [inline ntlm_test.go]: We really really don't need to pull in a new module for testing three variable values.

File: ntlm.go (hunk 5 above)
Comment (sboutzen) [reply ntlm.go]: I agree it is kind of a generic name, but the name of the function is chosen as a reference to the related part of the specification. Does that change your mind about the function name?

File: ntlm.go (hunk 8 above)
Comment (sboutzen) [reply ntlm.go]: I added references to official microsoft documentation, and an unofficial walk through which imho explains everything well. Also i added a reference to the specific place in the unofficial walk through that touches upon the contents of LMv2 and NTLMv2, in the spot where i create those payloads.

File: ntlm.go (hunk 10 above)
Comment (sboutzen) [reply ntlm.go]: You're right, the purpose of the encapsulation can be easily achieved with a couple of comments on the blocks that construct the two payloads.

File: ntlm.go (hunk 11 above)
Comment (sboutzen) [reply ntlm.go]: I refactored the structure of everything around this spot to make it more readable. Hopefully i succeeded! (:

File: ntlm_test.go (hunk 13 above)
Comment (sboutzen) [reply ntlm_test.go]: True

**Comment by sboutzen** [comment]
Running the tests locally does not result in DATA RACE, so im not sure why CI fails

**Comment by kardianos** [comment]
Try running test with `go test -v -race -cpu 4 -count 10` locally, ensuring you have a connection env var dsn defined.

**Comment by sboutzen** [comment]
I ran all the tests again against my local database, 10 are skipped (only on windows) and the rest are passing. Also, the build failed after i made changes that do not impact anything.

**Comment by sboutzen** [comment]
I am able to occasionally reproduce the data race, will take a look into what is going on.

**Comment by kardianos** [comment]
@sboutzen I will not gate the PR on resolving this, as you say, it may happen on master as well.

**Comment by sboutzen** [comment]
@kardianos Are you waiting for something from me? :)

**Pull request merged**